File: cs373/2/java/TestPrimes.java
// setenv CLASSPATH .:/lusr/share/lib/java/junit/junit-4.5.jar
// javac -Xlint TestPrimes.java
// java -ea TestPrimes

import java.io.IOException;
import java.io.StringWriter;
import java.io.Writer;
import java.util.Arrays;
import java.util.Scanner;
import junit.framework.Assert;
import junit.framework.TestCase;
import junit.framework.TestSuite;
import junit.textui.TestRunner;

public class TestPrimes extends TestCase {
	public static void main (String[] args) {
		TestRunner.run(new TestSuite(TestPrimes.class));
	}

	// testRead ------------------------------------------------------------

	// Can it deal with typical input?
	public void testRead1 () {
		Scanner scanner = new Scanner("8");
		int[] value = new int[1];
		boolean result = Primes.read(scanner, value);

		Assert.assertTrue(result == true);
		Assert.assertTrue(value[0] == 8);
	}

	// Can it deal with zeros?
	public void testRead2 () {
		Scanner scanner = new Scanner("0");
		int[] value = new int[1];
		boolean result = Primes.read(scanner, value);

		Assert.assertTrue(result == true);
		Assert.assertTrue(value[0] == 0);
	}

	// Can it deal with huge numbers?
	public void testRead3 () {
		Scanner scanner = new Scanner("" + Integer.MAX_VALUE);
		int[] value = new int[1];
		boolean result = Primes.read(scanner, value);

		Assert.assertTrue(result == true);
		Assert.assertTrue(value[0] == Integer.MAX_VALUE);
	}

	// Can it deal with negative numbers?
	public void testRead4 () {
		Scanner scanner = new Scanner("-1");
		int[] value = new int[1];
		boolean result = Primes.read(scanner, value);

		Assert.assertTrue(result == true);
		Assert.assertTrue(value[0] == -1);
	}

	// Can it deal with white space?
	public void testRead5 () {
		Scanner scanner = new Scanner(" \n\t1 \n\t");
		int[] value = new int[1];
		boolean result = Primes.read(scanner, value);

		Assert.assertTrue(result == true);
		Assert.assertTrue(value[0] == 1);
	}

	// testPrint -----------------------------------------------------------

	// Can it deal with typical output?
	public void testPrint1 () throws IOException {
		Writer writer = new StringWriter();
		int[] values = new int[] {2, 3, 5, 7};
		Primes.print(writer, values);
		String result = writer.toString();

		Assert.assertTrue(result.equals("2 3 5 7\n"));
	}

	// Can it deal with huge numbers?
	public void testPrint2 () throws IOException {
		Writer writer = new StringWriter();
		int[] values = new int[] {Integer.MAX_VALUE, 3, 5, 7};
		Primes.print(writer, values);
		String result = writer.toString();

		Assert.assertTrue(result.equals(Integer.MAX_VALUE + " 3 5 7\n"));
	}

	// Can it deal with impossible output?
	public void testPrint3 () throws IOException {
		Writer writer = new StringWriter();
		int[] values = new int[] {};
		Primes.print(writer, values);
		String result = writer.toString();

		Assert.assertTrue(result.equals("Impossible.\n"));
	}

	// testEval ------------------------------------------------------------
	
	// Can it deal with typical input?
	public void testEval1 () {
		int value = 8;
		int[] result = Primes.eval(value);
		int[] correct = new int[] {2, 2, 2, 2};

		Assert.assertTrue(Arrays.equals(result, correct));
	}
	
	// Can it deal with huge input?
	public void testEval2 () {
		int value = 10000000;
		int[] result = Primes.eval(value);
		int[] correct = new int[] {2, 2, 5, 9999991};

		Assert.assertTrue(Arrays.equals(result, correct));
	}
	
	// Can it deal with input that fails?
	public void testEval3 () {
		int value = 7;
		int[] result = Primes.eval(value);
		int[] correct = new int[] {};

		Assert.assertTrue(Arrays.equals(result, correct));
	}
	
	// Can it deal with negative input that fails?
	public void testEval4 () {
		int value = -1;
		int[] result = Primes.eval(value);
		int[] correct = new int[] {};

		Assert.assertTrue(Arrays.equals(result, correct));
	}

	// testIsPrime ---------------------------------------------------------
	
	// Can it deal with typical prime input?
	public void testIsPrime1 () {
		int value = 2;
		boolean result = Primes.isPrime(value);

		Assert.assertTrue(result);
	}
	
	// Can it deal with typical composite input?
	public void testIsPrime2 () {
		int value = 4;
		boolean result = Primes.isPrime(value);

		Assert.assertFalse(result);
	}
}

File: cs373/2/python/TestPrimes.py
#!/lusr/bin/python

# python TestPrimes.py

import StringIO
import unittest

from Primes import primes_read, primes_print, primes_eval, primes_is_prime

class TestPrimes (unittest.TestCase):
	# test_read ------------------------------------------------------------

	# Can it deal with typical input?
	def test_read1 (self):
		reader = StringIO.StringIO('8\n')
		value = [0]
		result = primes_read(reader, value)
		self.assert_(result == True)
		self.assert_(value[0] == 8)

	# Can it deal with zeros?
	def test_read2 (self):
		reader = StringIO.StringIO('0\n')
		value = [0]
		result = primes_read(reader, value)
		self.assert_(result == True)
		self.assert_(value[0] == 0)

	# Can it deal with huge numbers?
	def test_read3 (self):
		reader = StringIO.StringIO('10000000\n')
		value = [0]
		result = primes_read(reader, value)
		self.assert_(result == True)
		self.assert_(value[0] == 10000000)

	# Can it deal with negative numbers?
	def test_read4 (self):
		reader = StringIO.StringIO('-1\n')
		value = [0]
		result = primes_read(reader, value)
		self.assert_(result == True)
		self.assert_(value[0] == -1)

	# Can it deal with white space?
	def test_read5 (self):
		reader = StringIO.StringIO(' \t1 \t\n')
		value = [0]
		result = primes_read(reader, value)
		self.assert_(result == True)
		self.assert_(value[0] == 1)

	# test_print -----------------------------------------------------------

	# Can it deal with typical input?
	def test_print1 (self):
		writer = StringIO.StringIO()
		primes_print(writer, [2, 3, 5, 7])
		self.assert_(writer.getvalue() == '2 3 5 7\n')

	# Can it deal with huge numbers?
	def test_print2 (self):
		writer = StringIO.StringIO()
		primes_print(writer, [10000000, 3, 5, 7])
		self.assert_(writer.getvalue() == '10000000 3 5 7\n')

	# Can it deal with impossible input?
	def test_print3 (self):
		writer = StringIO.StringIO()
		primes_print(writer, [])
		self.assert_(writer.getvalue() == 'Impossible.\n')

	# test_eval ------------------------------------------------------------

	# Can it deal with typical input?
	def test_eval1 (self):
		primes = primes_eval(17)
		self.assert_(primes == [2, 3, 5, 7])

	# Can it deal with huge numbers?
	def test_eval2 (self):
		primes = primes_eval(10000000)
		self.assert_(primes == [2, 2, 5, 9999991])

	# Can it deal with input that fails?
	def test_eval3 (self):
		primes = primes_eval(7)
		self.assert_(primes == [])

	# Can it deal with negative input?
	def test_eval4 (self):
		primes = primes_eval(-1)
		self.assert_(primes == [])

	# test_is_prime --------------------------------------------------------

	# Can it deal with typical prime input?
	def test_is_prime1 (self):
		self.assert_(primes_is_prime(2) == True)

	# Can it deal with typical composite input?
	def test_is_prime2 (self):
		self.assert_(primes_is_prime(4) == False)

if __name__ == '__main__':
	unittest.main()
